### PR TITLE
[4.0] Plugin as services

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Loads the plugins through the boot functionality similar to what we have for modules and components.

A plugin needs then a file services/provider.php which does register the `PluginInterface::class` resource in the container:
```
public function register(Container $container)
{
	$container->set(
		PluginInterface::class,
		function (Container $container)
		{
			// @Todo This needs to be changed to a proper factory
			$plugin = \Joomla\CMS\Plugin\PluginHelper::getPlugin('content', 'myplugin');

			return new MyPlugin(
				$container->get(DispatcherInterface::class),
				(array)$plugin
			);
		}
	);
}
```

More information can be found in #19834, #20217 and #19667.

_This is on hold till #19834 got merged as it is based on it._

### Testing Instructions
- Log in on the back end.
- Open back end admin panel end.

### Expected result
The joomla update check will be executed.

### Actual result
All is working fine.

### Documentation Changes Required
Needs to be documented like the component service locator stuff.

### CODE REVIEW

**Comment by wilsonge** [comment]
I'm going to make a decision to reject this. All our plugins are basically singletons - they are booted once and then that instance is used for the entire request cycle - unlike components and modules. The booting is done transparently to everyone using the CMS. I think in this case introducing a dispatcher is complete overkill. There are definitely optimisations we can make (making the plugin config array an object so that we can build the plugins optionally through the DIC) and obviously using the namespace stored in the extensions table in the DB so they can be namespaced. But I don't think a dispatcher is the way to go here

**Comment by laoneo** [comment]
We do not have a dispatcher here at all. But it makes sense to have the same procedure for every extension to be loaded the same way. It would be a logical next step to get rid of the file path lookup thing for plugins too.

**Comment by mbabker** [comment]
> But it makes sense to have the same procedure for every extension to be loaded the same way.

No, it doesn't.  You can't apply this same logic to language packages or templates.  Different extension types have different purposes, they can't all be loaded and managed in the same way.

Components are service providers.  Modules really should just be component tasks in a HMVC based environment.  Both of those are in essence task executors, it makes sense to have some form of dispatching mechanism in front of them.  Plugins are intended to be an event subscriber.  They don't need a dispatching mechanism in front of them, and as long as it is mandated they extend the CMSPlugin class you can't efficiently use dependency injection with them.  If we can deal with that subclassing issue then maybe it makes sense to support the service provider file to be able to do dependency injection, otherwise I kind of agree that the dispatcher concept here is overkill.

**Comment by laoneo** [comment]
Just to be clear here, there is no dispatcher involved for plugins. It is just about instantion for plugins which is now in the hands of the plugin itself. The dispatcher stuff is from the module pr where this pr is built on top of it. When the module pr gets merged then this one will get much smaler. Thats why I labeled it "On hold".

**Comment by laoneo** [comment]
Reopening as the module service https://github.com/joomla/joomla-cms/pull/19834 got merged.

File: libraries/src/Extension/PluginInterface.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Access to plugin specific services.
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+interface PluginInterface extends DispatcherAwareInterface
+{
+	/**
+	 * Registers it's listeners.
```
Comment (Quy) [inline libraries/src/Extension/PluginInterface.php]:
```suggestion
	 * Registers its listeners.
```

File: libraries/src/Extension/ExtensionManagerTrait.php [hunk 2]
```diff
@@ -167,6 +193,50 @@ private function loadExtension($type, $extensionName, $extensionPath)
 		return $extension;
 	}
 
+	/**
+	 * Creates a CMS plugin from the filesystem.
+	 *
+	 * @param   string  $plugin  The plugin
+	 * @param   string  $type    The type
+	 *
+	 * @return CMSPlugin
```
Comment (Quy) [inline libraries/src/Extension/ExtensionManagerTrait.php]:
```suggestion
	 * @return  CMSPlugin
```

**Comment by wilsonge** [comment]
I honestly still don't really agree with this for the reasons mentioned earlier. But as per our discussion on glip earlier merged.

**Pull request merged**